src/peft/tuners/vblora/layer.py

```diff
@@ -0,0 +1,263 @@
+                        raise ValueError(
+                            f"NaNs detected in the merged weights. The adapter {active_adapter} seems to be broken"
+                        )
+
+                    base_layer.weight.data = orig_weights
+                else:
+                    base_layer.weight.data += self.get_delta_weight(active_adapter)
+                self.merged_adapters.append(active_adapter)
+
+    def unmerge(self) -> None:
+        if not self.merged:
+            warnings.warn("Already unmerged. Nothing to do.")
+            return
+
+        while len(self.merged_adapters) > 0:
+            active_adapter = self.merged_adapters.pop()
+            if active_adapter in self.vblora_logits_A.keys():
+                self.get_base_layer().weight.data -= self.get_delta_weight(active_adapter)
+
+    @staticmethod
+    def _get_low_rank_matrix(logits, vblora_vector_bank, topk):
+        top_k_logits, indices = logits.topk(topk, dim=-1)
+        topk_weights = F.softmax(top_k_logits, dim=-1)
+        return (topk_weights.unsqueeze(-1) * vblora_vector_bank[indices]).sum(-2)
+
+    def get_delta_weight(self, adapter) -> torch.Tensor:
+        """
+        Compute the delta weight for the given adapter.
+
+        Args:
+            adapter (str):
+                The name of the adapter for which the delta weight should be computed.
+        """
+        vblora_logits_A = self.vblora_logits_A[adapter]
+        vblora_logits_B = self.vblora_logits_B[adapter]
+        vblora_vector_bank = self.vblora_vector_bank[adapter]
+
+        device = vblora_logits_A.device
+        dtype = vblora_logits_A.dtype
+
+        # In case users wants to merge the adapter weights that are in
+        # float16 while being on CPU, we need to cast the weights to float32, perform the merge and then cast back to
+        # float16 because the `@` and matmul operation in general is not supported in torch + cpu + fp16.
+        cast_to_fp32 = device.type == "cpu" and dtype == torch.float16
+        if cast_to_fp32:
+            vblora_logits_A = vblora_logits_A.float()
+            vblora_logits_B = vblora_logits_B.float()
+            vblora_vector_bank = vblora_vector_bank.float()
+        # A: (rank, in_tile, vector_length) -> (rank, in_tile x vector_length)
+        A = Linear._get_low_rank_matrix(vblora_logits_A, vblora_vector_bank, self.topk[adapter]).reshape(
+            vblora_logits_A.shape[0], -1
+        )
+        # B: (out_tile, rank, vector_length) -> (out_tile, vector_length, rank) -> (out_tile x vector_length, rank)
+        B = (
+            Linear._get_low_rank_matrix(vblora_logits_B, vblora_vector_bank, self.topk[adapter])
+            .transpose(1, 2)
+            .reshape(-1, vblora_logits_B.shape[1])
+        )
+
+        output_tensor = transpose(B @ A, self.fan_in_fan_out)
+        return output_tensor
+
+    def forward(self, x: torch.Tensor, *args, **kwargs) -> torch.Tensor:
+        previous_dtype = x.dtype
```

Comment:
I wonder if we can make a sanity check here that users are not trying to train a model that was loaded with `save_topk_weights`. E.g. we could check if `self.training` and if there is -inf in the logits, WDYT? If this check is expensive, it could be done only during the first iteration.

Reply:
I added some lines of code to check if both `self.training` and `save_topk_weights` are true.

Reply:
I added a sanity check based on your feedback.